webdav_storage/src/WebDav.php

```diff
@@ -64,36 +68,96 @@ private function pathForRef(string $reference): array
 		$fold2 = substr($reference, 2, 2);
 		$file  = substr($reference, 4);
 
-		return [$this->encodePath(implode('/', [$fold1, $fold2, $file])), implode([$fold1, $fold2]), $file];
+		return [$this->encodePath(implode('/', [$fold1, $fold2, $file])), implode('/', [$fold1, $fold2]), $file];
 	}
 
 	/**
 	 * URL encodes the given path but keeps the slashes
 	 *
 	 * @param string $path to encode
+	 *
 	 * @return string encoded path
 	 */
-	protected function encodePath($path): string
+	protected function encodePath(string $path): string
 	{
 		// slashes need to stay
 		return str_replace('%2F', '/', rawurlencode($path));
 	}
 
-	protected function exists($uri): bool
+	protected function exists(string $uri): bool
 	{
 		return $this->client->head($uri, [HTTPClientOptions::AUTH => $this->authOptions])->getReturnCode() == 200;
 	}
 
-	protected function mkcol($uri)
+	protected function hasItems(string $uri): bool
+	{
+		$dom               = new \DOMDocument('1.0', 'UTF-8');
+		$dom->formatOutput = true;
+		$root              = $dom->createElementNS('DAV:', 'd:propfind');
+		$prop              = $dom->createElement('d:allprop');
+
+		$dom->appendChild($root)->appendChild($prop);
+
+		$opts = [
+			HTTPClientOptions::AUTH    => $this->authOptions,
+			HTTPClientOptions::HEADERS => ['Depth' => 1],
+			HTTPClientOptions::BODY    => $dom->saveXML(),
+		];
+
+		$response = $this->client->request('propfind', $uri, $opts);
+		$xml      = XML::parseString($response->getBody());
+
+		print_r($xml);
+
+		return true;
+	}
+
+	protected function mkcol($uri): bool
 	{
-		return $this->client->request('mkcol', $uri, [HTTPClientOptions::AUTH => $this->authOptions])->getReturnCode() == 200;
+		return $this->client->request('mkcol', $uri, [HTTPClientOptions::AUTH => $this->authOptions])
+							->getReturnCode() == 200;
+	}
+
+	protected function checkAndCreatePath(string $fullPath): void
+	{
+		$finalUrl = $this->url . '/' . trim($fullPath, '/');
+
+		if ($this->exists($finalUrl)) {
+			return;
+		}
+
+		$pathParts = explode('/', trim($fullPath, '/'));
+		$path      = '';
+
+		foreach ($pathParts as $part) {
+			$path .= '/' . $part;
+			$partUrl = $this->url . $path;
+			if (!$this->exists($partUrl)) {
+				$this->mkcol($partUrl);
+				;
+			}
+		}
+	}
+
+	protected function checkAndDeletePath(string $fullPath): void
```

Comment:
What is this about?

Reply:
I thought we should avoid empty directories after deleting the last item of it, so after a delete, we scan recursively the directories and delete them if no item is left..

during typing, I got another idea: Maybe for performance reason, we could add it per bulk to a Worker instead :D

Reply:
Ah, I asked before of the call you later removed in `put()`. It makes slightly more sense now but it still doesn't perform any deletion call?

Reply:
nope, it's the last point before I'm ready.
The crux is that this is the most complex part.

All other functionality can be done with given HTTP methods, provided by `HTTPClient`. The "check" method in that case needs to select the items of the webserver first, which now needs to parse and understand the webdav XML-multistatus response, which is way more logic than the rest so far :D